### PR TITLE
[CARBONDATA-2504][STREAM] Support StreamSQL for streaming job

### PR DESCRIPTION
**Pull request by jackylk**
Currently, user need to write Spark Streaming APP to use carbon streaming ingest feature, which is not so easy for some users. By providing StreamSQL, user can manage the streaming job more easily.

## StreamSQL

```
    sql(
      s"""
         |CREATE TABLE source(
         | id INT,
         | name STRING,
         | city STRING,
         | salary FLOAT,
         | tax DECIMAL(8,2),
         | percent double,
         | birthday DATE,
         | register TIMESTAMP,
         | updated TIMESTAMP
         |)
         |TBLPROPERTIES (
         | 'streaming'='source',
         | 'format'='csv',
         | 'path'='$csvDataDir'
         |)
      """.stripMargin)

    sql(
      s"""
         |CREATE TABLE sink(
         | id INT,
         | name STRING,
         | city STRING,
         | salary FLOAT,
         | tax DECIMAL(8,2),
         | percent double,
         | birthday DATE,
         | register TIMESTAMP,
         | updated TIMESTAMP
         |)
         |TBLPROPERTIES (
         |  'streaming'='sink'
         |)
      """.stripMargin)

    sql(
      """
        |CREATE STREAM stream123 ON TABLE sink
        |TBLPROPERTIES(
        |  'trigger'='ProcessingTime',
        |  'interval'='1 seconds')
        |AS
        |  SELECT *
        |  FROM source
        |  WHERE id % 2 = 1
      """.stripMargin)

    sql("DROP STREAM stream123")

    sql("SHOW STREAMS [ON TABLE sink]")
```

   SHOW STREAMS will print

   | Stream name              | status  | Source | Sink | start time          | time elapsed |
   | ------------------ | ------- | ------ | ---- | ------------------- | ------------ |
   | stream123 | Started | device | fact | 2018-02-03 14:32:42 | 10d2h32m     |

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import org.apache.spark.sql.streaming.StreamingQuery
+
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable
+import org.apache.carbondata.spark.StreamingOption
+import org.apache.carbondata.streaming.CarbonStreamException
+
+object StreamJobManager {
+  private val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+  private val jobs = mutable.Map[String, StreamJobDesc]()
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: Need to use concurrent hash map

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonShowStreamsCommand.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      Row(
+        job.streamingQuery.id.toString,
+        if (job.streamingQuery.isActive) "RUNNING" else "FAILED",
+        s"${ job.sourceDb }.${ job.sourceTable }",
+        s"${ job.sinkDb }.${ job.sinkTable }",
+        new Date(job.startTime).toString,
+        String.format(
+          "%s days, %s hours, %s min, %s sec",
+          TimeUnit.MILLISECONDS.toDays(elapsedTime).toString,
+          TimeUnit.MILLISECONDS.toHours(elapsedTime).toString,
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonShowStreamsCommand.scala]: toHours will give total hours elapsed
toMinutres will give total minutes elapsed.
It will not in format (10days, 1hours, 5min, 34 sec)

**Comment by xuchuanyin** [comment]
Since you provide `CREATE STREAM SOURCE XXX` grammar, do we need another `SHOW STREAM SOURCE ...`?

**Comment by xuchuanyin** [comment]
In the description of PR, better to optimize the output of `SHOW STREAMS` and `Kill STREAM` and use the tableNames you provided in the above `Start a streaming job`.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6019/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4861/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5031/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5033/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5037/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5038/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonRelation.scala [hunk 3]
```diff
@@ -22,6 +22,7 @@ import scala.Array.canBuildFrom
 import scala.collection.JavaConverters._
 import scala.util.parsing.combinator.RegexParsers
 
+import org.apache.spark.sql.CarbonEnv
```
Comment (xuchuanyin) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonRelation.scala]: ?
Only add an import without changing the code?

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      sinkTable: CarbonTable,
+      query: String,
+      streamDf: DataFrame,
+      options: StreamingOption): String = {
+    val latch = new CountDownLatch(1)
+    var exception: Throwable = null
+    var job: StreamingQuery = null
+
+    // start a new thread to run the streaming ingest job, the job will be running
+    // until user stops it by STOP STREAM JOB
```
Comment (xuchuanyin) [inline integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: outdated `STOP STREAM JOB`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    val tableModel = new CarbonSpark2SqlParser().prepareTableModel(
+      ifNotExistPresent = false,
+      dbName,
+      tableName,
+      fields,
+      Seq.empty,
+      mutable.Map[String, String](tblProperties.toSeq: _*),
+      None
+    )
+    val tableInfo = TableNewProcessor.apply(tableModel)
```
Comment (xuchuanyin) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala]: The `Create Stream Source` internally calls `Create Table`. What's the difference between them? If there is no difference, is `Create Stream Source` necessary? Can ordinary carbon table can be used as source?

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 6]
```diff
@@ -145,6 +149,55 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
...
+      case dbName ~ tableName ~ options ~ query =>
+        val optionMap = options.getOrElse(List[(String, String)]()).toMap[String, String]
+        CarbonCreateStreamCommand(dbName, tableName, optionMap, query)
+    }
+
+  /**
+   * The syntax of KILL STREAM
+   * KILL STREAM ON TABLE [dbName].tableName
+   */
+  protected lazy val killStream: Parser[LogicalPlan] =
```
Comment (xuchuanyin) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: What will happen if I call `Kill Stream`?
Should I (re)`create stream` if I want the stream start again? If so, better to change the grammar to `DROP STREAM ON TABLE ...`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * This command is used to create Stream Source, which is implemented as a Carbon Table
+ */
+case class CarbonCreateStreamSourceCommand(
+    dbName: Option[String],
+    tableName: String,
+    fields: Seq[Field],
+    tblProperties: Map[String, String]
+) extends MetadataCommand {
+  override def processMetadata(sparkSession: SparkSession): Seq[Row] = {
+    val tableModel = new CarbonSpark2SqlParser().prepareTableModel(
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala]: 1) 'streaming'='source' need not be explicitly passed by user, for this ddl can internally always add.
2) select * on stream src Table should fail.
3) another ddl required to show stream source tables. As it is tough to identify streaming source tables from list of tables.
4) Stream sink table DDL mentioned in description not required as carbondata already supports streaming ingestion.

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala (hunk 1 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonRelation.scala (hunk 3 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonRelation.scala]: I will remove it

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonShowStreamsCommand.scala (hunk 2 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonShowStreamsCommand.scala]: ok

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala (hunk 4 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala (hunk 5 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala]: Create Stream Source is removed

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 6 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: If the stream is dropped, user need to trigger CREATE STREAM again

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala (hunk 7 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamSourceCommand.scala]: 1. ok

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6304/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5141/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5263/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6313/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5271/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5151/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6316/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5154/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5272/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6320/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5158/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6321/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5277/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5159/

File: integration/spark-common/src/main/java/org/apache/carbondata/spark/util/SparkDataTypeConverterImpl.java [hunk 8]
```diff
@@ -171,4 +176,21 @@ public Object wrapWithGenericRow(Object[] fields) {
     }
     return fields;
   }
+
+  public static StructType convertToSparkSchema(ColumnSchema[] carbonColumns) {
```
Comment (ravipesala) [inline integration/spark-common/src/main/java/org/apache/carbondata/spark/util/SparkDataTypeConverterImpl.java]: I think this method do not handle complex datatype schema. Can you check `CarbonRelation` how we are handling complex schema and same way can be used here

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.apache.spark.sql.streaming.{ProcessingTime, Trigger}
+
+import org.apache.carbondata.common.exceptions.sql.MalformedCarbonCommandException
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.path.CarbonTablePath
+import org.apache.carbondata.streaming.parser.CarbonStreamParser
+
+class StreamingOption(val userInputMap: Map[String, String]) {
+  def trigger: Trigger = {
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala]: Ideally, these all could be lazy val instead of def

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    val trigger = userInputMap.getOrElse(
+      "trigger", throw new MalformedCarbonCommandException("trigger must be specified"))
+    val interval = userInputMap.getOrElse(
+      "interval", throw new MalformedCarbonCommandException("interval must be specified"))
+    trigger match {
+      case "ProcessingTime" => ProcessingTime(interval)
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala]: `ProcessingTime is deprecated, suggested using `Trigger.ProcessingTime`

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+import org.apache.carbondata.common.exceptions.NoSuchStreamException
+import org.apache.carbondata.common.exceptions.sql.MalformedCarbonCommandException
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable
+import org.apache.carbondata.spark.StreamingOption
+import org.apache.carbondata.spark.util.CarbonScalaUtil
+import org.apache.carbondata.streaming.CarbonStreamException
+import org.apache.carbondata.streaming.parser.CarbonStreamParser
+
+object StreamJobManager {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: Add description to the class. And also better mention that this stream job is only available to the driver memory and not persisted so other drivers cannot see ongoing stream jobs.

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+  private val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+
+  // map of stream name to job desc
+  private val jobs = new ConcurrentHashMap[String, StreamJobDesc]()
+
+  private def validateStreamName(streamName: String): Unit = {
+    if (StreamJobManager.getAllJobs.exists(_.streamName.equalsIgnoreCase(streamName))) {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: I guess just checking `jobs.containsKey(streamName)` should be enough.

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 13]
```diff
@@ -145,6 +149,41 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
...
+        val optionMap = options.getOrElse(List[(String, String)]()).toMap[String, String]
+        CarbonCreateStreamCommand(streamName, dbName, tableName, optionMap, query)
+    }
+
+  /**
+   * The syntax of DROP STREAM
+   * DROP STREAM streamName
+   */
+  protected lazy val dropStream: Parser[LogicalPlan] =
+    DROP ~> STREAM ~> ident <~ opt(";") ^^ {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Better support `if exist` also

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 14]
```diff
@@ -145,6 +149,41 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
...
     }
 
+  /**
+   * The syntax of CREATE STREAM
+   * CREATE STREAM streamName ON TABLE [dbName.]tableName
+   * [STMPROPERTIES('KEY'='VALUE')]
+   * AS SELECT COUNT(COL1) FROM tableName
+   */
+  protected lazy val createStream: Parser[LogicalPlan] =
+    (CREATE ~> STREAM ~> ident) ~ (ON ~> TABLE ~> (ident <~ ".").?) ~ ident ~
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: Better support `if not exist` also

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamCommand.scala [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    var sourceTable: CarbonTable = null
+
+    // find the streaming source table in the query
+    // and replace it with StreamingRelation
+    val streamLp = df.logicalPlan transform {
+      case r: LogicalRelation
+        if r.relation.isInstanceOf[CarbonDatasourceHadoopRelation] &&
+           r.relation.asInstanceOf[CarbonDatasourceHadoopRelation].carbonTable.isStreamingSource =>
+        val (source, streamingRelation) = prepareStreamingRelation(sparkSession, r)
+        if (sourceTable != null && sourceTable.getTableName != source.getTableName) {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamCommand.scala]: I don't get the logic here, You want to verify that more than one should not be present here? Then better update the error to `Stream query on more than one stream source table is not supported`

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6352/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5190/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5303/

File: integration/spark-common/src/main/java/org/apache/carbondata/spark/util/SparkDataTypeConverterImpl.java (hunk 8 above)
Comment (jackylk) [reply integration/spark-common/src/main/java/org/apache/carbondata/spark/util/SparkDataTypeConverterImpl.java]: fixed

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala (hunk 9 above)
Comment (jackylk) [reply integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala]: fixed

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala (hunk 10 above)
Comment (jackylk) [reply integration/spark-common/src/main/scala/org/apache/carbondata/spark/StreamingOption.scala]: but it is still there in spark 2.1.0

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala (hunk 12 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala (hunk 11 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/carbondata/stream/StreamJobManager.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 13 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 14 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamCommand.scala (hunk 15 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/stream/CarbonCreateStreamCommand.scala]: fixed

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5310/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6361/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5199/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6367/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5205/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5316/

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala [hunk 16]
```diff
@@ -123,12 +123,9 @@ class TestStreamingTableOperation extends QueryTest with BeforeAndAfterAll {
 
     createTable(tableName = "agg_table", streaming = true, withBatchLoad = false)
 
-    var csvDataDir = new File("target/csvdatanew").getCanonicalPath
+    val csvDataDir = integrationPath + "target/csvdatanew"
```
Comment (ravipesala) [inline integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: it should be `integrationPath + "/spark2/target/csvdatanew`

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala [hunk 17]
```diff
@@ -123,12 +123,9 @@ class TestStreamingTableOperation extends QueryTest with BeforeAndAfterAll {
...
     generateCSVDataFile(spark, idStart = 10, rowNums = 5, csvDataDir)
     generateCSVDataFile(spark, idStart = 10, rowNums = 5, csvDataDir, SaveMode.Append)
-    csvDataDir = new File("target/csvdata").getCanonicalPath
-    // streaming ingest 10 rows
-    generateCSVDataFile(spark, idStart = 10, rowNums = 10, csvDataDir)
```
Comment (ravipesala) [inline integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: why is this removed?

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala [hunk 18]
```diff
@@ -195,7 +192,6 @@ class TestStreamingTableOperation extends QueryTest with BeforeAndAfterAll {
     sql("USE default")
     sql("DROP DATABASE IF EXISTS streaming CASCADE")
     new File("target/csvdatanew").delete()
```
Comment (ravipesala) [inline integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: it should be integrationPath + "/spark2/target/csvdatanew

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala [hunk 19]
```diff
@@ -1642,7 +1638,9 @@ class TestStreamingTableOperation extends QueryTest with BeforeAndAfterAll {
     var rows = sql("SHOW STREAMS").collect()
     assertResult(0)(rows.length)
 
-    val csvDataDir = new File("target/csvdata").getCanonicalPath
+    val csvDataDir = integrationPath + "/target/csvdata"
```
Comment (ravipesala) [inline integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: it should be integrationPath + "/spark2/target/csvdatanew

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala (hunk 17 above)
Comment (jackylk) [reply integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: I added it in individual test case

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala (hunk 18 above)
Comment (jackylk) [reply integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: fixed

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala (hunk 19 above)
Comment (jackylk) [reply integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: fixed

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala (hunk 16 above)
Comment (jackylk) [reply integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: fixed

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6383/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5218/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6390/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5224/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5331/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5337/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6404/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5238/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5348/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6420/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5257/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6426/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5361/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5362/

**Comment by ravipesala** [comment]
LGTM
